packages/blocks/src/page-block/edgeless/components.ts

```diff
@@ -188,3 +166,147 @@ export function EdgelessBlockChildrenContainer(
+    let handles: TemplateResult | null = null;
+    if (this.state.type === 'none') return handles;
+    if (!this.state.active) {
+      const leftCenter = [
+        rect.x,
+        rect.y + rect.height / 2 + (PADDING_Y * this.zoom) / 2,
+      ];
+      const rightCenter = [
+        rect.x + rect.width + PADDING_X * this.zoom,
+        rect.y + rect.height / 2 + (PADDING_Y * this.zoom) / 2,
+      ];
+      const handleLeft = Handle(
+        leftCenter[0],
+        leftCenter[1],
+        'left',
+        this._onHandleMouseDown
+      );
+      const handleRight = Handle(
+        rightCenter[0],
+        rightCenter[1],
+        'right',
+        this._onHandleMouseDown
+      );
+      handles = html` ${handleLeft}${handleRight} `;
+    }
+    return handles;
+  }
+
+  private _onHandleMouseDown = (e: MouseEvent, type: 'left' | 'right') => {
+    // prevent selection action being fired
+    e.stopPropagation();
+    if (this.state?.type !== 'none') {
+      const {
+        rect,
+        selected: { xywh },
+      } = this.state;
+      const [x] = JSON.parse(xywh) as XYWH;
+      this._dragStartInfo = {
+        startMouseLeft: e.clientX,
+        absoluteLeft: x,
+        // the width of the selected group may 0 after init use rect.width instead
+        width: rect.width,
+        direction: type,
+      };
+      // parent ele is the edgeless block container
+      this.parentElement?.addEventListener('mousemove', this._onDrag);
+      this.parentElement?.addEventListener('mouseup', this._onDragEnd);
+    }
+  };
+
+  private _onDrag = (e: MouseEvent) => {
+    let newX = 0;
+    let newW = 0;
+    if (this.state.type !== 'none') {
+      const {
+        selected,
+        selected: { space, xywh },
+      } = this.state;
+      const [x, y, w] = JSON.parse(xywh) as XYWH;
+      const minus = this._dragStartInfo.startMouseLeft - e.clientX;
+      if (this._dragStartInfo.direction === 'left') {
+        newX = this._dragStartInfo.absoluteLeft - minus / this.zoom;
+        newW = (this._dragStartInfo.width + minus) / this.zoom;
+      } else {
+        newX = x;
+        newW = (this._dragStartInfo.width - minus) / this.zoom;
+      }
+      // limit the width of the selected group
+      if (newW < GROUP_MIN_LENGTH) {
+        newW = GROUP_MIN_LENGTH;
+        newX = x;
+      }
+      // if xywh do not change, no need to update
+      if (newW === w && newX === x) {
+        return;
+      }
+      const groupBlock = getBlockById<'div'>(selected.id);
+      const groupContainer = groupBlock?.parentElement;
+      // first change container`s x/w directly for get groups real height
+      if (groupContainer) {
+        groupContainer.style.width = newW + 'px';
+        groupContainer.style.translate = `translate(${newX}px, ${y}px) scale(${this.zoom})`;
+      }
+      // reset the width of the container may trigger animation
+      requestAnimationFrame(() => {
+        // refresh xywh by model
+        space.updateBlock(selected, {
```

Comment:
We need to call `store.captureSync()` once before the first time the drag move handler gets triggered. This can prevent the drag operation from being merged with previous operations, ensuring the current operation can be undone standalone.